### DISCUSSION TITLE
Using other tts models exported to onnxruntime with sherpa-onnx

### DISCUSSION THREAD

**Discussion by tn-17** [k2-fsa/sherpa-onnx#890] [General]
Besides VITS, can other tts models based on VITS and are exported to onnxruntime work with sherpa-onnx?
* https://github.com/p0p4k/vits2_pytorch
* https://github.com/fishaudio/Bert-VITS2
* https://github.com/RVC-Boss/GPT-SoVITS/tree/main

If they are not compatible out of the box, what needs to be done to make them compatible? I am interesting in learning how to implement support for those models for the sherpa-onnx on python.

**Comment by csukuangfj** [comment]
The first step is to convert them to onnx.

**Comment by yaambe** [comment]
I successfully exported a VITS model to ONNX after fine-tuning an MMS model using Hugging Face tools. You can find the tools I used here: https://github.com/ylacombe/finetune-hf-vits.

However, I'm having trouble finding documentation on how to convert these VITS models to Sherpa ONNX. Any guidance or resources would be greatly appreciated!

**Reply by csukuangfj** [reply]
We support converting MMS to sherpa-onnx!

Please see
https://k2-fsa.github.io/sherpa/onnx/tts/mms.html

<img width="1129" alt="Screenshot 2024-06-27 at 18 21 09" src="https://github.com/k2-fsa/sherpa-onnx/assets/5284924/5f25772d-060f-4b22-92f5-73712ec45a9e">

**Reply by yaambe** [reply]
Thank you for your response. 
However, the model I am working with is a Transformers VITS model, not MMS VITS model.
I have reviewed the documentation at https://k2-fsa.github.io/sherpa/onnx/tts/mms.html, but it specifically covers MMS models.

**Reply by csukuangfj** [reply]
we don't support transformer.

**Reply by yaambe** [reply]
In case somebody wants to use a transformers vits model with sherpa-onnx. Here is a script to convert to convert the model such that its compatible with sherpa-onnx.
```
import torch
from transformers import VitsModel
import onnx
from typing import Dict, Any

# Load the model
model_name = "/path/to/transformers/vits/model" 
model = VitsModel.from_pretrained(model_name).eval()

class VitsWrapper(torch.nn.Module):
    def __init__(self, model: VitsModel):
        super().__init__()
        self.model = model
        self.use_sdp = model.config.use_stochastic_duration_prediction

    def forward(
        self,
        x,
        x_length,
        noise_scale=0.667,
        length_scale=1.0,
        noise_scale_w=0.8,
    ):        
        
        if self.use_sdp:
            self.model.noise_scale = noise_scale_w
        else:
            self.model.noise_scale = noise_scale   
                     
        self.model.speaking_rate = 1.0 / length_scale    
        
        # Use x_length and noise_scale in a way that doesn't affect the output
        dummy_tensor = torch.zeros(1, device=x.device)
        dummy_tensor = dummy_tensor + x_length.sum() * 0.0 + noise_scale * 0.0

        output = self.model(
            input_ids=x,     
            ).waveform
        
        # Add the dummy tensor to the output without changing its value
        output = output + dummy_tensor * 0.0
        
        return output

wrapped_model = VitsWrapper(model)

x = torch.randint(low=1, high=10, size=(50,), dtype=torch.int64)

x = x.unsqueeze(0)
x_length = torch.tensor([x.shape[1]], dtype=torch.int64)
noise_scale = torch.tensor([1], dtype=torch.float32)
length_scale = torch.tensor([1], dtype=torch.float32)
noise_scale_w = torch.tensor([1], dtype=torch.float32)

exported_model_name = "model_sherpa.onnx"

# Export the traced model to ONNX
torch.onnx.export(
    wrapped_model,
    (x, x_length, noise_scale, length_scale, noise_scale_w),
    exported_model_name,
    opset_version=13,
    input_names=[
        "x",
        "x_length",
        "noise_scale",
        "length_scale",
        "noise_scale_w",
    ],
    output_names=['y'],
    dynamic_axes={
    "x": {0: "N", 1: "L"},
    "x_length": {0: "N"},
    "y": {0: "N", 2: "L"},
    }
)

print(f"Model exported to {exported_model_name}")

def add_meta_data(filename: str, meta_data: Dict[str, Any]):
    """Add meta data"""
    model = onnx.load(filename)
    for key, value in meta_data.items():
        meta = model.metadata_props.add()
        meta.key = key
        meta.value = str(value)

    onnx.save(model, filename)

# Define the metadata
meta_data = {
    "model_type": "vits",
    "comment": "mms",
    "add_blank": 1, 
    "language": "div", # mms language code here
    "frontend": "characters",
    "n_speakers": 0,
    "sample_rate": 16000,
}

# Specify the path to your ONNX model
onnx_model_path = exported_model_name

# Add metadata to the model
add_meta_data(onnx_model_path, meta_data)

print(f"Metadata added to {onnx_model_path}")

# Optionally, verify the metadata was added
model = onnx.load(onnx_model_path)
print("\nUpdated model metadata:")
for prop in model.metadata_props:
    print(f"{prop.key}: {prop.value}")

# verify all input is there
input_names = [input.name for input in model.graph.input]    
print("\ninputs:", input_names)
```

**Reply by csukuangfj** [reply]
Thank you!

Would you mind uploading the exported model to huggingface?

**Reply by yaambe** [reply]
Check out the following link. 
[https://huggingface.co/yaambe/mms-tts-spa/tree/main](https://huggingface.co/yaambe/mms-tts-spa/tree/main)
Its a Spanish model.

**Reply by thewh1teagle** [reply]
Thank you so much! it worked for me with other mms models. I also needed to prepare `vocab.json` into tokens.txt but that's simple

**Reply by thewh1teagle** [reply]
@csukuangfj 
I also exported the Hebrew model to hugging face, maybe it's useful [mms-tts-heb](https://huggingface.co/thewh1teagle/mms-tts-heb/tree/main)

**Reply by willwade** [reply]
> Thank you!
> 
> Would you mind uploading the exported model to huggingface?

Is there a neater way than me publishing them on my own huggingface account? Like a pr on a onnx-mms account or something? I’ll invest time in converting them but I feel I’d be best not hosting them 


This way too the community know what’s been converted and hasn’t.

**Comment by willwade** [comment]
I’m currently converting all mms models at https://huggingface.co/willwade/mms-tts-multilingual-models-onnx/tree/main

they are going up via a big automated script. 
code here https://github.com/willwade/mms-tts-multilingual-models-onnx

I’m creating wav files when I can
